### PR TITLE
Implemented argpartition #18938

### PR DESCRIPTION
**Pull request by Aayushkumr**

### CODE REVIEW

File: ivy/functional/backends/torch/sorting.py [hunk 1]
```diff
@@ -61,6 +61,18 @@ def msort(
...
+# argpartition
+@with_unsupported_dtypes({"2.0.1 and below": ("complex",)}, backend_version)
+def argpartition(
+    x: torch.Tensor,
+    /,
+    kth: int,
+    *,
+    axis: int = -1,
+) -> torch.Tensor:
+    return torch.argpartition(x, kth=kth, dim=axis)
```
Comment (xoiga123) [inline ivy/functional/backends/torch/sorting.py:73]: `torch.argpartition` doesn't exist?

File: ivy/functional/backends/tensorflow/sorting.py [hunk 2]
```diff
@@ -61,6 +61,22 @@ def msort(
...
+)
+def argpartition(
+    x: paddle.Tensor,
+    /,
+    kth: Union[int, paddle.Tensor],
+    *,
+    axis: int = -1,
+    order: Optional[str] = None,
+) -> paddle.Tensor:
+    return paddle.argsort(x, axis=axis).argsort(axis=axis)
```
Comment (xoiga123) [inline ivy/functional/backends/tensorflow/sorting.py:77]: This should be a `tensorflow` function, not `paddle` :sweat_smile:

File: ivy/functional/backends/paddle/sorting.py [hunk 3]
```diff
@@ -122,3 +122,18 @@ def msort(
...
+)
+def argpartition(
+    x: paddle.Tensor,
+    /,
+    kth: Union[int, paddle.Tensor],
+    *,
+    axis: int = -1,
+    order: Optional[str] = None,
+) -> paddle.Tensor:
+    return paddle.argsort(x, axis=axis).argsort(axis=axis)
```
Comment (xoiga123) [inline ivy/functional/backends/paddle/sorting.py:139]: `argpartition` and `argsort` is not the same, [reference](https://www.pythonpool.com/numpy-argpartition/)

File: ivy/functional/frontends/jax/numpy/searching_sorting.py [hunk 4]
```diff
@@ -61,6 +61,39 @@ def msort(a):
...
+    },
+    "jax",
+)
+def argpartition(
+    a,
+    kth,
+    axis=-1,
+    kind="stable",
+    order=None,
+) -> ivy.Array:
```
Comment (xoiga123) [inline ivy/functional/frontends/jax/numpy/searching_sorting.py:80]: Correct me if I'm wrong but the signature for `jax.numpy.argpartition` doesn't have `kind` and `order` arguments? [reference](https://jax.readthedocs.io/en/latest/_autosummary/jax.numpy.argpartition.html)

File: ivy/functional/frontends/jax/numpy/searching_sorting.py [hunk 5]
```diff
@@ -61,6 +61,39 @@ def msort(a):
...
+        )
+    if order is not None:
+        raise ivy.utils.exceptions.IvyError(
+            "'order' argument to argpartition is not supported."
+        )
+
+    return ivy.argpartition(a, kth, axis=axis)
+
+
+argpartition.support_native_out = True
```
Comment (xoiga123) [inline ivy/functional/frontends/jax/numpy/searching_sorting.py:94]: I don't think it supports an `out` argument natively, [reference](https://jax.readthedocs.io/en/latest/_autosummary/jax.numpy.argpartition.html)

File: ivy/functional/frontends/jax/numpy/searching_sorting.py (hunk 4 above)
Comment (xoiga123) [inline ivy/functional/frontends/jax/numpy/searching_sorting.py:80]: We used the decorator `@to_ivy_arrays_and_back`, which means that the return type of the function should be a frontend array, not `ivy.Array`. It's best not to typehint the frontend implementations.

File: ivy/functional/frontends/numpy/sorting_searching_counting/sorting.py [hunk 6]
```diff
@@ -25,6 +25,30 @@ def msort(a):
...
+            ] = -1
+            partition_indices[
+                ivy.where(
+                    partition_indices == -1,
+                    ivy.expand_dims(k, axis=axis),
+                    partition_indices,
+                )
+            ]
+        partitions.append(partition_indices)
+    return ivy.stack(partitions, axis=axis)
```
Comment (xoiga123) [inline ivy/functional/frontends/numpy/sorting_searching_counting/sorting.py:49]: Why didn't you use `ivy.argpartition`?

File: ivy/functional/frontends/torch/comparison_ops.py [hunk 7]
```diff
@@ -253,6 +253,33 @@ def msort(input, *, out=None):
...
+            ] = -1
+            partition_indices[
+                ivy.where(
+                    partition_indices == -1,
+                    ivy.expand_dims(k, axis=axis),
+                    partition_indices,
+                )
+            ]
+        partitions.append(partition_indices)
+    return ivy.stack(partitions, axis=axis)
```
Comment (xoiga123) [inline ivy/functional/frontends/torch/comparison_ops.py:280]: Why didn't you use `ivy.argpartition`?

File: ivy/functional/ivy/sorting.py [hunk 8]
```diff
@@ -363,3 +363,108 @@ def searchsorted(
         out=out,
         ret_dtype=ret_dtype,
     )
+
+
+@handle_exceptions
+@handle_nestable
+@handle_array_like_without_promotion
+@handle_out_argument
```
Comment (xoiga123) [inline ivy/functional/ivy/sorting.py:371]: Why did you use `@handle_out_argument` when the function doesn't support `out`?

File: ivy/functional/ivy/sorting.py [hunk 9]
```diff
@@ -363,3 +363,108 @@ def searchsorted(
...
+@to_native_arrays_and_back
+@handle_array_function
+@handle_device_shifting
+def argpartition(
+    x: Union[ivy.Array, ivy.NativeArray],
+    kth: Union[int, List[int]],
+    /,
+    *,
+    axis: int = -1,
+    kind: Optional[str] = "introselect",
```
Comment (xoiga123) [inline ivy/functional/ivy/sorting.py:381]:
```python
kind: str = "introselect"
```
We don't use `Optional` typehint when the default value is not `None`

File: ivy/functional/ivy/sorting.py [hunk 10]
```diff
@@ -363,3 +363,108 @@ def searchsorted(
...
+    Returns
+    -------
+    ret
+        An array of indices that index the data in `x` along the given `axis` in
+        partitioned order.
+
+    This function conforms to the `Array API Standard
+    <https://data-apis.org/array-api/latest/>`_. This docstring is an extension of
+    the `docstring <https://data-apis.org/array-api/latest/
+    API_specification/generated/array_api.argpartition.html>`_
```
Comment (xoiga123) [inline ivy/functional/ivy/sorting.py:425]: Is the link correct? I got an `404` trying to access it.

File: ivy_tests/test_ivy/test_frontends/test_jax/test_numpy/test_searching_sorting.py [hunk 11]
```diff
@@ -468,3 +468,38 @@ def test_jax_unique(fn_inputs, frontend, test_flags, fn_tree, on_device):
         return_counts=return_counts,
         axis=axis,
     )
+
+
+# argpartition
+@handle_frontend_test(
+    fn_tree="jax.numpy.argpartition",
+    dtype_x_kth_axis=helpers.dtype_values_kth_axis(
```
Comment (xoiga123) [inline ivy_tests/test_ivy/test_frontends/test_jax/test_numpy/test_searching_sorting.py:476]: `helpers.dtype_values_kth_axis` doesn't exist in the codebase?

File: ivy_tests/test_ivy/test_frontends/test_torch/test_comparison_ops.py [hunk 12]
```diff
@@ -762,3 +762,74 @@ def test_torch_topk(
...
+    helpers.test_frontend_function(
+        input_dtypes=input_dtype,
+        frontend=frontend,
+        test_flags=test_flags,
+        fn_tree=fn_tree,
+        on_device=on_device,
+        input=input[0],
+        dim=axis,
+        k=k,
+    )
```
Comment (xoiga123) [inline ivy_tests/test_ivy/test_frontends/test_torch/test_comparison_ops.py:799]: Is this mistakenly repeated :sweat_smile:?

**Review by xoiga123 — CHANGES_REQUESTED** [review]
Sorry it took so long to review! I have some changes that you may want to look at. Thank you for contributing! :hugs:

**Comment by ivy-seed** [comment]
This PR has been labelled as stale because it has been inactive for more than 7 days. If you would like to continue working on this PR, then please add another comment or this PR will be closed in 7 days.

**Comment by zhumakhan** [comment]
Hi, thanks for you contribution! Could you please go through each comment left by xoiga123 and resolve them?

**Comment by ivy-seed** [comment]
This PR has been closed because it has been marked as stale for more than 7 days with no activity.
